src/org/rascalmpl/repl/completers/RascalCommandCompletion.java

```diff
@@ -0,0 +1,84 @@
+package org.rascalmpl.repl.completers;
+
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.SortedSet;
+import java.util.TreeMap;
+import java.util.function.BiConsumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.jline.reader.Candidate;
+import org.jline.reader.Completer;
+import org.jline.reader.LineReader;
+import org.jline.reader.ParsedLine;
+import org.rascalmpl.interpreter.utils.StringUtils;
+import org.rascalmpl.interpreter.utils.StringUtils.OffsetLengthTerm;
+public class RascalCommandCompletion implements Completer {
+    private static final NavigableMap<String,String> COMMAND_KEYWORDS;
+    static {
+        COMMAND_KEYWORDS = new TreeMap<>();
+        COMMAND_KEYWORDS.put("set", "change a evaluator setting");
+        COMMAND_KEYWORDS.put("undeclare", "undeclare a local variable of the REPL");
+        COMMAND_KEYWORDS.put("help", "print help message");
+        COMMAND_KEYWORDS.put("edit", "open a rascal module in the editor");
+        COMMAND_KEYWORDS.put("unimport", "unload an imported module from the REPL");
+        COMMAND_KEYWORDS.put("declarations", "show declarations"); // TODO figure out what it does
+        COMMAND_KEYWORDS.put("quit", "cleanly exit the REPL");
+        COMMAND_KEYWORDS.put("history", "history"); // TODO: figure out what it does
+        COMMAND_KEYWORDS.put("test", "run rest modules");
+        COMMAND_KEYWORDS.put("modules", "show imported modules");// TODO: figure out what it does 
+        COMMAND_KEYWORDS.put("clear", "clear repl screen");
+    }
+
+    private final NavigableMap<String, String> setOptions;
+    private final BiConsumer<String, List<Candidate>> completeIdentifier;
+    private final BiConsumer<String, List<Candidate>> completeModule;
+    public RascalCommandCompletion(NavigableMap<String, String> setOptions, BiConsumer<String, List<Candidate>> completeIdentifier, BiConsumer<String, List<Candidate>> completeModule) {
```

Comment:
In the end yes, and since this is called frequently, I thought to keep the design of jline3 (namely: here, append onto this list) around. I've not measured it if concatenation would be an acceptable performance tradeoff.